### PR TITLE
BUG: sample IDs that look like scientific notation are treated as numbers

### LINKED ISSUE #130: BUG: ANCOM-BC Fails if sample IDs look too much like exponent
**Bug Description**
We had a user on the forum that reported getting the error message "duplicate 'row.names' are not allowed " from ANCOM-BC. For this user it was because their sample-ids looked like exponents. My theory is that the sample-ids are losing some resolution when they are converted to exponents and so they are becoming "duplicates". I was able to fix this by replacing sample-ids with E to F and then it worked fine. 

[Forum x-ref](https://forum.qiime2.org/t/how-to-make-barplots-for-selected-features/28899/14)

**Steps to reproduce the behavior**
1. Run ancom-bc with forum users data

**Steps to reproduce the solution**
In python: 
1.  from qiime2 import Artifact
2. import pandas as pd
3. table = Artifact.load("/Users/chloeherman/Downloads/forum-table.qza")
4. df = table.view(pd.DataFrame)
5.  df.to_csv("/Users/chloeherman/Downloads/forum-table-3.txt", sep = "\t", header = True)
6. edit ids in txt to not have "E"
7. Go to R 
8. run read.delim(file, check.names = FALSE, row.names = 1)  < this will fail if you haven't done step 6.

### PR DESCRIPTION
**Pull request by colinbrislawn**
Trying to close #130

More testing is needed, especially a regression test for sampleIDs like this.

### CODE REVIEW

**Comment by lizgehret** [comment]
Thanks for working on this @colinbrislawn! Lmk when it's ready for a review and I'll take a look 🙂

**Comment by colinbrislawn** [comment]
This seems to work:
```R 
t(read.delim(inp_abundances_path,
  check.names = TRUE, # change to true to so the empty first column becomes .$X
  row.names = 1,
  colClasses = c(X = "character") # Make the .$X column a character, leave other columns
))
```

Any idea why `check.names = F` originally?

For the regression test, maybe import this file and make sure the sample names match?
[table-SI-names.tsv](https://github.com/qiime2/q2-composition/pull/131/files#diff-462e6f113b779494e37f20eb030276b4051683398cbb59a05112e04041aafd26)

Could I just run `test_ids_in_table_not_in_md()` again, with a new set of files?

**Comment by lizgehret** [comment]
> This seems to work:
> 
> ```r
> t(read.delim(inp_abundances_path,
>   check.names = TRUE, # change to true to so the empty first column becomes .$X
>   row.names = 1,
>   colClasses = c(X = "character") # Make the .$X column a character, leave other columns
> ))
> ```
> 
> Any idea why `check.names = F` originally?

There wasn't a strong reason for that being set to False in the initial implementation - I hadn't run into this issue while developing, so I didn't realize this would be an edge case we'd need to consider. This looks reasonable to me though!

> For the regression test, maybe import this file and make sure the sample names match? [table-SI-names.tsv](https://github.com/qiime2/q2-composition/pull/131/files#diff-462e6f113b779494e37f20eb030276b4051683398cbb59a05112e04041aafd26)
> 
> Could I just run `test_ids_in_table_not_in_md()` again, with a new set of files?

Yeah that should work! I think as long as we continue to use the original files/tests, and then just create additional tests with your new set of files to make sure the behavior is the same, that should work well.

**Comment by colinbrislawn** [comment]
Hey Liz, I think I'm ready for your help here.

I've got a new `test_ids_in_table_with_es` testing function that _should_ fail because I've reverted the ancom import code... but it works fine. I'm also unfamiliar with the dataloaf format

EDIT:
>and then just create additional tests with your new set of files to make sure the behavior is the same, that should work well

Yes! I've added those files. Can you help me write that?

**Comment by lizgehret** [comment]
Yes, happy to help! Hung up with a few other things this afternoon, but I'll take a closer look tomorrow 🙂

File: q2_composition/assets/run_ancombc.R [hunk 1]
```diff
@@ -69,8 +69,9 @@ output_loaf         <- opt$output_loaf
 if (!file.exists(inp_abundances_path)) {
   errQuit("Input file path does not exist.")
 } else {
-  otu_file <- t(read.delim(inp_abundances_path, check.names = FALSE,
-                            row.names = 1))
+  # This is the old code, which should fail with the new test data
+  otu_file <- t(read.delim(inp_abundances_path, check.names = FALSE, # TRUE,
+                            row.names = 1)) #, colClasses = c(X = "character")))
```
Comment (colinbrislawn) [inline q2_composition/assets/run_ancombc.R]: Once we have a unit test that fails to import table-ancom-IDE.qza,
we can fix this function with
`check.names = TRUE`
`row.names = 1, colClasses = c(X = "character")))`

File: q2_composition/tests/test_ancombc.py [hunk 2]
```diff
@@ -50,9 +55,26 @@ def test_ids_in_table_not_in_md(self):
         with self.assertRaisesRegex(KeyError, 'Not all samples present within'
                                     ' the table were found in the associated'
                                     ' metadata file.*L6S68'):
-            ancombc(table=self.table, metadata=self.missing_md,
+            ancombc(table=self.table, metadata=self.md_missing_md,
                     formula='bodysite')
 
+    # TODO: IDs with Es that look like scientific notation 004e002
+    def test_ids_in_table_with_es(self):
```
Comment (colinbrislawn) [inline q2_composition/tests/test_ancombc.py]: I just copied a function below as a placeholder. This is checking for columns, while we are worried about sampleIDs

**Comment by lizgehret** [comment]
Okay just adding some notes here for myself as I'm investigating:

- Forum user who reported this issue was using QIIME 2 2023.9 and R 4.2.2.
- I replicated the error with the same package versions using their dataset.
- I created a dummy table/md file inline in a test to replicate the behavior using the same sample IDs that the forum user did to make sure the error would catch (which it did).
- I added two tests (that both fail) for uppercase E and lowercase E (to ensure that R doesn't care about casing).

Adding just these two tests (not finalized) in a commit so you can take a look @colinbrislawn!

**_EDIT_**: Let me know if you want me to finish this up, or if you'd rather take it across the finish line - I'm happy either way, just don't want to steal your thunder!

**Comment by lizgehret** [comment]
Another thing I decided to do @colinbrislawn was to just create the table/MD files inline instead of using your newly created files - sorry for flip flopping on that, I realized that would be easier for me to see/iterate on the table as I was testing!

**Comment by colinbrislawn** [comment]
Okay, I'll take a look. I'm excited to see how you set up testing.

>Let me know if you want me to finish this up, or if you'd rather take it across the finish line

I would like more experience with Python, so I'll see if I can get this finished. This may be _more_ work for you, but if you are willing to help me out I would very much appreciate it.

**Comment by lizgehret** [comment]
> I would like more experience with Python, so I'll see if I can get this finished. This may be _more_ work for you, but if you are willing to help me out I would very much appreciate it.

Happy to help, and in full support of you learning! You just let me know what questions you have 🙂

**Comment by colinbrislawn** [comment]
I've run into issues around metadata, so I've updated both import functions.
(Good thing we tested!)

In the metadata importe, I reference the sampleID column with the key `sample-id`.
Is this first column always called that in this plugin or should I support other names?

Zooming out a little, Tidyverse packages like readr may give us better ways to handle this.
We could start the tidyverse retooling, or save it for the ancom-bc2 function!

**Comment by lizgehret** [comment]
Hey @colinbrislawn, `sample-id` is only one of the supported names for metadata identifiers (that's just the one I happened to use in the original test data); the full list can be found [here](https://docs.qiime2.org/2023.9/tutorials/metadata/#identifier-column). I'm a bit wary of hard-coding all of these into the import function unless absolutely necessary - but definitely open to any improvements that might be available within tidyverse, since that's already a dependency!

Re: ancom-bc2 - we were waiting for that paper to be published before starting development on this method, but it looks like that's finally happened as of a little over a month ago! I'll chat with the rest of the eng team in this week's meeting about this and see what thoughts are there.

**Comment by colinbrislawn** [comment]
I thought I remembered a bunch of allowed `sample-id`s!

Here's the options I can see:
- Use the python code to make sure that column is called `sample-id` before passing it to the R code I have in this PR
- Switch to the tidyverse `readr::read_delim()` and use their better API.

What do you suggest?

**Comment by lizgehret** [comment]
> I thought I remembered a bunch of allowed `sample-id`s!
> 
> Here's the options I can see:
> 
>     * Use the python code to make sure that column is called `sample-id` before passing it to the R code I have in this PR
> 
>     * Switch to the tidyverse `readr::read_delim()` and use their better API.
> 
> 
> What do you suggest?

Let's see what can be done with `readr::read_delim()` - I was reviewing the available parameters for this method in the [R docs](https://www.rdocumentation.org/packages/readr/versions/2.1.5/topics/read_delim), and I'm wondering if a good solution would be to utilize the `col_types` param and set that to `character` for the first column (i.e. the index column containing the sample IDs). I haven't investigated this in detail, so not sure if this will actually work like we want it to, but it seems like a good option to explore!

**Comment by lizgehret** [comment]
Hey @colinbrislawn - just as a head's up, I'm going to have to bump this one to the next release cycle (as our current planned release is _technically_ today - but realistically it's looking like it'll be tomorrow/Friday). We can definitely get this in as a patch to the current release epoch though, so these changes wouldn't have to wait an entire cycle before being added into the ecosystem.

Let me know if you need anything from me or have any questions!

**Comment by colinbrislawn** [comment]
Let's push it to the next release or try a patch.

Tidyverse should be easy to implement, but I'm away from my main dev PC right now so I'm a little delayed...

**Comment by lizgehret** [comment]
Sounds good @colinbrislawn, just let me know if you need anything from me or when you're ready for another review 🙂

**Comment by gregcaporaso** [comment]
@colinbrislawn, feel free to ping me for a review too (@lizgehret is out for the next few weeks).

**Comment by colinbrislawn** [comment]
Thanks Greg. Metadata has proven as hard as expected, but I've got feature data parsing up and running.

The `inp_abundances_path` now gets parsed into an `otu_file` with `readr::read_tsv()`, and the first column is always treated as a string. I think this is good to go!

I tried to import `inp_metadata_path` in the same way, but downstream steps need a base R data.frame with rownames, and Tidyverse does not believe in those. So I'm sticking with base-R for now.

I've got one more bug to fix before we merge...

File: q2_composition/assets/run_ancombc.R [hunk 3]
```diff
@@ -69,20 +69,48 @@ output_loaf         <- opt$output_loaf
...
   }
 
 if (!file.exists(inp_metadata_path)) {
   errQuit("Metadata file path does not exist.")
 } else {
-  metadata_file <- read.delim(inp_metadata_path, check.names = FALSE,
-                              row.names = 1)
+  md_file <- read.delim(inp_metadata_path,
+                        check.names = FALSE, row.names = 1,
+                        colClasses = list(`sample-id` = "character"))
```
Comment (colinbrislawn) [inline q2_composition/assets/run_ancombc.R]: @gregcaporaso 

This updated function supports sampleIDs like `1e002`

But it only worked by selecting a column called `sample-id`
So `sampleIDs ` and `sample-ids` would not work.

Can we fix this upstream with or do we have to support all valid first column names here?

Comment (gregcaporaso) [reply q2_composition/assets/run_ancombc.R]: @colinbrislawn, I'm not certain. If the name of that column is normalized before it gets into R, then you should be set (and that would be the ideal way to handle it, so that we don't have to embed knowledge of the allowed column names here in addition to in the metadata parsing logic - i.e., the DRY principle). Have you tried passing the same metadata in to the QIIME 2 action with different column names? If you could try that and it works, you're all set and don't need to do it here. If it doesn't work, we should address that upstream (you may be able to rename that column before it's passed to R in the ANCOMBC action).

Comment (colinbrislawn) [reply q2_composition/assets/run_ancombc.R]: Oh, that's a great idea.
I'll try this later today and report back!

Comment (colinbrislawn) [reply q2_composition/assets/run_ancombc.R]: Okay,

This new function works for typical sample IDs and any `SampleID/sample id/#SampleID` column name.

However, it only can fix SI notation IDs (100e2) when the column is named `sample-id` due to how I identify this column.

>the name of that column is normalized before it gets into R

Yes, I'll see if I can get this added into this line:
https://github.com/qiime2/q2-composition/blob/dev/q2_composition/_ancombc.py#L89

**Comment by colinbrislawn** [comment]
Checks pass on my machine!

I've got to remove my own notes and code before we merge...

**Comment by colinbrislawn** [comment]
@lizgehret, I've got the R code working, including a patch to _ancombc.py to always output `sample-id`

When you have a chance, could you take a look at the unit test?

Thanks!

**Comment by lizgehret** [comment]
Hey @colinbrislawn!

Sorry for the delay, still getting caught up on things after vacation and travel. Will give this a proper review shortly!

**Review by lizgehret — COMMENTED** [review]
Hey @colinbrislawn,

Sorry (again) for the delay on this! Overall this looks reasonable to me, thanks again for working on this! I think the remaining task here is (just as you mention in your test comment) to add assertions into the contents of the dataloaf itself for these new test cases.

One additional (small) comment is that we may want to re-name IDE with something a bit more verbose (like IDs-with-Es, or something like that). Only reason I bring that up is because I was initially thinking of IDE in terms of interactive development environment and that may cause confusion in the future.

**Comment by colinbrislawn** [comment]
>I was initially thinking of IDE in terms of interactive development environment

Good idea. I noticed that too when I wrote it. This I can fix.

> add assertions into the contents of the dataloaf itself for these new test cases.

I'm unfamiliar with the dataloaf format. Would you like me to try this first or is this something you could wrap up for me?

**Comment by lizgehret** [comment]
> I'm unfamiliar with the dataloaf format. Would you like me to try this first or is this something you could wrap up for me?

I can finish that up, thanks @colinbrislawn! I'll wait for your name change updates re: IDE and then take it from there. 🙂

**Comment by lizgehret** [comment]
okay @colinbrislawn i've updated these tests to be more comprehensive - they now assert that the contents of the resultant IDs in each slice of the dataloaf are what we'd expect.

i also apologize for asking you to update the names of the files you added that contained IDE - i realized after the fact that we didn't need those files after all 🤦‍♀️ 

i'm going to get someone else to give this a final review (@colinvwood TIA!) since i'm no longer an unbiased reviewer 😉 but this should be g2g!

File: q2_composition/assets/run_ancombc.R [hunk 4]
```diff
@@ -69,20 +69,28 @@ output_loaf         <- opt$output_loaf
 if (!file.exists(inp_abundances_path)) {
   errQuit("Input file path does not exist.")
 } else {
-  otu_file <- t(read.delim(inp_abundances_path, check.names = FALSE,
-                            row.names = 1))
+  # Tidyverse
+  otu_file <- inp_abundances_path |>
+    # The sample IDs are in the first column, which we treat as characters.
+    # All other columns are feature frequencies as doubles
+    read_tsv(col_names = T, col_types = cols("character", .default = col_double())) |>
```
Comment (colinvwood) [inline q2_composition/assets/run_ancombc.R]:
```suggestion
    read_tsv(col_names = TRUE, col_types = cols("character", .default = col_double())) |>
```

File: q2_composition/assets/run_ancombc.R [hunk 5]
```diff
@@ -69,20 +69,28 @@ output_loaf         <- opt$output_loaf
...
   }
 
 if (!file.exists(inp_metadata_path)) {
   errQuit("Metadata file path does not exist.")
 } else {
-  metadata_file <- read.delim(inp_metadata_path, check.names = FALSE,
-                              row.names = 1)
+  md_file <- read.delim(inp_metadata_path,
+                        check.names = FALSE, row.names = 1,
+                        colClasses = list(`sample-id` = "character"))
```
Comment (colinvwood) [inline q2_composition/assets/run_ancombc.R]: Should we maybe use either tidyverse importing funcs or the base ones (but not both) for consistency? Then in the future a maintainer wouldn't need to research both `col_types` and `colClasses` for example.

Comment (lizgehret) [reply q2_composition/assets/run_ancombc.R]: ah yeah that's probably a good call @colinvwood, thanks for pointing that out. i'll test out updating md_file to use tidyverse and see if that still does what we want it to
